src/Is/ip.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * @since 0.0.2
+ */
+import { isString } from "./string";
+
+/**
+ * Validate IPv4
+ * This is a modified version of github.com/chriso/validator.js `isIP` (version === 4)
+ *
+ * @since 0.0.2
+ * @category Is
+ */
+export const isIPv4 = <A>(val: A) => {
+  if (!val || !isString(val)) {
+    return false;
+  }
+
+  const _val = "" + val;
+
+  // It will accept anything with three dots (.) and numeric digits
+  const ipv4Regex = /^(\d+)\.(\d+)\.(\d+)\.(\d+)$/;
+  if (!ipv4Regex.test(_val)) {
+    return false;
+  }
+  // This part of code do checking for any invalid IP address like 256.256.256.256
+  const parts = _val.split(".").sort(function (a, b) {
+    return +a - +b;
+  });
+  return +parts[3] <= 255;
+};
+
+/**
+ * Validate IPv6
+ * This is a modified version of github.com/chriso/validator.js `isIP` (version === 6)
+ *
+ * @since 0.0.2
+ * @category Is
+ */
+export const isIPv6 = <A>(val: A) => {
+  if (!val || !isString(val)) {
+    return false;
+  }
+
+  const _val = "" + val;
+
+  const ipv6Block = /^[0-9A-F]{1,4}$/i;
+  const blocks = _val.split(":");
+  let foundOmissionBlock = false; // marker to indicate ::
+
+  // At least some OS accept the last 32 bits of an IPv6 address
+  // (i.e. 2 of the blocks) in IPv4 notation, and RFC 3493 says
+  // that '::ffff:a.b.c.d' is valid for IPv4-mapped IPv6 addresses,
+  // and '::a.b.c.d' is deprecated, but also valid.
+  const foundIPv4TransitionBlock = isIPv4(blocks[blocks.length - 1]);
+  const expectedNumberOfBlocks = foundIPv4TransitionBlock ? 7 : 8;
+
+  if (blocks.length > expectedNumberOfBlocks) return false;
+
+  // initial or final ::
+  if (_val === "::") {
+    return true;
+  } else if (_val.substr(0, 2) === "::") {
+    blocks.shift();
```

Comment:
can we use `Array.proptotype.reduce` instead of mutating the array using `pop()` and `shift()`